### ISSUE TITLE
Rule proposal: Improve object-curly-newline

### ISSUE THREAD

**Issue by Alaneor** [eslint/eslint#6369]
**What version of ESLint are you using?** 2.12

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```js
{ 'object-curly-newline': [1, {
      multiline: true,
      minProperties: 3
    }]
}
```

**What did you do? Please include the actual source code causing the issue.**

The rule warns for this snippet:

```js
const env = { // <- Unexpected a line break after this open brace
  NODE_ENV: 'test'
} // <- Unexpected a line break before this close brace
```

I would like to propose some kind of option to turn off warnings for situations where the total number of properties separated on new lines is lower than that of `minProperties`.

**Rationale**

I consider this rule a great safeguard against coding style such as this:
```js
const env = { first: 1, second: 2, third: 3, fourth: 4, /*...*/ }
```

But if the programmer feels that readability will be much better if the object property is put on its own line, even if it's the only property, I do not want that to cause a warning:
```js
const env = {
  first: fs.readFileSync('./some/maybe/a/bit/longer/string.txt')
}
```

This is currently not possible with the given set of options for this rule.

**Possible solution**

Introduce a third configuration parameter, `strictMinProperties` (bool, default: `true`), which, when set to `false`, would ignore cases where the total number of properties on each line is less than `minProperties`.

The name of the parameter is definitely something to consider, `strictMinProperties` just sounds weird, but I fail to come up with anything better 😕 .

**Comment by mysticatea** [comment]
In fact, do you want [object-property-newline] rule?

[object-property-newline]: http://eslint.org/docs/rules/object-property-newline

**Comment by platinumazure** [comment]
Labeling as "question" for now since `object-property-newline` may cover the desired use case.

@Alaneor Please take a look at the `object-property-newline` rule and see if it meets your needs. If not, please clarify what you need changed for your use case.

Note that `object-curly-newline` and `object-property-newline` serve two different purposes and may need to be configured specifically in order to meet your needs. Please see this example below:

```js
var myObject = {     // <-- This newline is required or forbidden by object-curly-newline
    foo: "bar",      // <-- This newline is required or forbidden by object-property-newline
    baz: "bzzt"      // <-- This newline is required or forbidden by object-curly-newline
};
```

With object-curly-newline enabled to require newlines after object braces, and object-property-newline enabled to disallow newlines after properties, this would lead to the following style:

```js
var myObject = {
    foo: "bar", baz: "bzzt"    // All properties are on one line, but not on the same line as either object curly brace
};
```

Hope this helps-- again, please let us know if you still believe a rule change may be needed.

**Comment by Alaneor** [comment]
Thanks for all the suggestions. I will try to clarify my intents since the other proposed rule also does not quite serve my desired coding style.

I would like to allow the following three, and only these three, ways to declare objects:

**Single line, if the total number of properties does not exceed 2:**
```js
var obj = { prop1: 'a', prop2: 'b' }
```

**Forced multiline, if the total number of properties exceeds 2:**
```js
var obj = {
  prop1: 'a',
  prop2: 'b',
  prop3: 'c'
}
```

**Optional multiline, if the total number of properties is 2 or lower, at the sole discretion of the developer, if they think it enhances readability (or for any arbitrary reason):**
```js
var obj = {
  prop1: 'a'
}
```

### The problem

The first two I can achieve by a combination of `object-property-newline` and `object-curly-newline`. However, the last case is currently being flagged as a violation of `object-curly-newline` regardless of how I configure it - it forces me to write the object on a single line because it does not have 3 or more properties.

I would like `object-curly-newline` to enforce newline before/after brackets for 3 or more properties, but **ignore** objects which **do have newlines** before/after brackets for objects with 2 or less properties.

Does that make more sense? 😄

**Comment by mysticatea** [comment]
I'm sorry for too late. 
I got it. 

I made a proposal: https://github.com/eslint/eslint/issues/6488#issuecomment-227383081
The options of this rule are "conditions to require line breaks", I thought `either` option.

**Comment by nzakas** [comment]
Let's merge this discussion into #6488.
